### PR TITLE
Setup with Vagrant

### PR DESCRIPTION
**Pull request by tgqiang**
Vagrantfile and other related files have been added. I am tagging all of you for this request because I need to ensure that the Vagrant setup works for all of your machines.

NOTE: you need both [Vagrant](https://www.vagrantup.com/downloads.html) and [VirtualBox](https://www.virtualbox.org/wiki/Downloads) installed in your machines before proceeding.

Please look into `bootstrap.sh` in `vagrant_setup` directory and make changes to the shell commands regarding setting up the PostgreSQL database if required. When you are done, you can simply call `vagrant up` once - the setting up of the virtual machine and the provisioning (i.e. setting up of the development environment) will be done together. `vagrant up` only needs to be called once to set up the virtual machine. Should you make changes to the provisioning script `bootstrap.sh`, you can re-perform the provisioning by calling `vagrant reload`.

1. To enter the virtual machine, you can use `vagrant ssh`, no password is required.
2. When you are in the virtual machine, to exit, just call `logout` like you would in a Linux environment.
3. To shut down the virtual machine, call `vagrant halt`.
4. To destroy the virtual machine (aka **totally remove it from your system**), call `vagrant destroy`. Doing this means that you will have to call `vagrant up` again the next time if you need the virtual machine. Don't call `vagrant destroy` if there is no need to (and chances are you wouldn't really need to, unless the semester is over and you just want it out of the system).
5. To access the project files once you enter the virtual environment, do `cd /csmodify` and you should see the files there.
6. I assume most (if not all) of you are using 'postgres' user, with 'postgres' database and password '12345678' with port number '5432' (except @nlzz22 ; you will need to customize something somewhere else), meaning you probably won't need to make any changes to the `Vagrantfile` or the `bootstrap.sh` shell script in `/vagrant_setup`.
7. Running `python app.py` should work as normal and you should be able to access the app at `localhost:8080`. Note that invoking `nosetests` in the Linux virtual machine requires an additional flag `--exe` in order for `nosetests` to be able to collect and run the test suites (`nosetests --exe`). Running `nosetests` in Windows wouldn't need this additional flag.

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  # Every Vagrant development environment requires a box. You can search for
+  # boxes at https://atlas.hashicorp.com/search.
+  config.vm.box = "hashicorp/precise64"
+
+  # By default, running 'python app.py' will create a server for the app,
+  # which can be accessed at localhost:8080.
+  # 
+  # Therefore, We specify a guest port number that is not 80 to
+  # avoid conflicting with the Python app.
+  config.vm.network "forwarded_port", guest: 8080, host: 8080
```
Comment (nlzz22) [inline Vagrantfile]: Note: when we get the latest version of Vagrant which is Vagrant 1.9.3, there is a vagrant bug which fails if we do "vagrant up". See: [https://github.com/rapid7/metasploitable3/issues/106](https://github.com/rapid7/metasploitable3/issues/106) if you want to know what the error is.

The fix can be done by adding a host_ip parameter to the line above (line 14) in this file (Vagrantfile). It should look like this instead:
`config.vm.network "forwarded_port", guest: 8080, host: 8080, host_ip: "127.0.0.1"`

Consider updating the file or leave a comment about this somewhere.

**Comment by nlzz22** [comment]
I receive the following error message when I did a `vagrant up` : 
image

When I do `vagrant ssh`, I am not in the correct directory, I am in `/home/vagrant` directory, and inside this directory only contains one file `postinstall.sh`. To rectify this, I have to do `cd /vagrant` to change to the correct directory.

After running `python app.py` and going to my browser to view the page, I will receive the following error:
(probably due to database failing to connect)
image

File: Vagrantfile (hunk 1 above)
Comment (tgqiang) [inline Vagrantfile]: Will update the file, after getting feedback from the rest.

**Comment by tgqiang** [comment]
@nlzz22 Yes, you need to do `cd /vagrant` to get to the app files because that directory is where I set the files to appear in.

If I do `config.vm.synced_folder ".", "/home/vagrant"` then you will see the project files immediately when you enter the virtual environment, but then `vagrant ssh` will always ask for a password because the private key is then lost.

As for the error you are facing over PostgreSQL, it is because your PostgreSQL is running on port 5433 rather than the default port 5432. As for this, could you try adjusting your Vagrantfile accordingly?

**Comment by nlzz22** [comment]
> Yes, you need to do cd /vagrant to get to the app files because that directory is where I set the files to appear in.

In that case, please remember to include that in the instructions in future for the setup of vagrant.

> As for the error you are facing over PostgreSQL, it is because your PostgreSQL is running on port 5433 rather than the default port 5432. As for this, could you try adjusting your Vagrantfile accordingly?

I managed to get it fixed by changing the port number inside `local_database_data.py`  .
Note that vagrant will not work properly on a new device, which does not have `local_database_data.py`  file automatically created and properly configured variables to match that of `bootstrap.sh`

Otherwise, the vagrant setup is working properly and fine for me, thanks.

**Comment by tgqiang** [comment]
@nlzz22 @helloqx @a0129998 @xaterz Just to inform you all that I've updated the PR's description, in case you've read the old description already.

**Comment by a0129998** [comment]
note: user must enable virtualization on their machine

**Pull request merged**